modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPartitionTimestampCursor.java

```diff
@@ -65,36 +61,32 @@ public boolean hasNext() {
 
             createVersionChainCursorIfMissing();
 
-            currentChain = null;
+            currentRowId = null;
 
             while (true) {
-                if (!versionChainCursor.hasNext()) {
+                if (!cursor.hasNext()) {
                     iterationExhausted = true;
 
                     return false;
                 }
 
-                VersionChain chain = versionChainCursor.next();
-
-                ReadResult result = readResultByRowId.remove(chain.rowId());
+                ReadResult result = cursor.next();
 
-                if (result == null) {
-                    // TODO: IGNITE-18717 Add lock by rowId
-                    chain = storage.readVersionChain(chain.rowId());
+                RowId rowId = result.rowId();
 
-                    if (chain == null) {
-                        continue;
-                    }
-
-                    result = findRowVersion(chain);
+                if (result.isEmpty()) {
```

Comment:
Total lack of comments makes this code difficult to read and understand.
For some reason, my guts tell me that this code might be problematic, but I don't understand what it does :(

Reply:
Try to fix it